File: Problems/Algorithms/1658. Minimum Operations to Reduce X to Zero/MinOperations.kt
class Solution {
    fun minOperations(nums: IntArray, x: Int): Int {
        var sum = 0
        for (num in nums) {
            sum += num
        }
        
        var len = -1
        var curr = 0
        var j = 0
        for (i in 0..nums.size-1) {
            curr += nums[i]
            while (j <= i && curr > sum - x) {
                curr -= nums[j]
                j++
            }
            
            if (curr == sum - x) {
                len = maxOf(len, i-j+1)
            }
        }
        
        if (len == -1) return -1
        return nums.size - len
    }
}

File: Problems/Algorithms/1658. Minimum Operations to Reduce X to Zero/min-operations.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int minOperations(vector<int>& nums, int x) {
        int target = 0;
        for (int num : nums) {
            target += num;
        }
        target -= x;

        int n = nums.size();
        if (target == 0) {
            return n;
        }

        int ans = 0;
        int curr = 0, left = 0;
        for (int right = 0; right < n; right++) {
            curr += nums[right];
            while (left <= right && curr > target) {
                curr -= nums[left];
                left++;
            }

            if (curr == target) {
                ans = max(ans, right - left + 1);
            }
        }

        return ans ? n - ans : -1;
    }
};
